File: deob-ast/src/main/java/dev/openrs2/deob/ast/transform/NewInstanceTransformer.java
package dev.openrs2.deob.ast.transform;

import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.Node;
import com.github.javaparser.ast.expr.MethodCallExpr;
import dev.openrs2.deob.ast.util.NodeUtils;

public final class NewInstanceTransformer extends Transformer {
	@Override
	public void transform(CompilationUnit unit) {
		NodeUtils.walk(unit, Node.TreeTraversal.POSTORDER, MethodCallExpr.class, expr -> {
			if (!expr.getNameAsString().equals("newInstance")) {
				return;
			}

			expr.getScope().ifPresent(scope -> {
				if (scope.isMethodCallExpr() && scope.asMethodCallExpr().getNameAsString().equals("getConstructor")) {
					return;
				}

				expr.setScope(new MethodCallExpr(scope.clone(), "getDeclaredConstructor"));
			});
		});
	}
}

File: deob-ast/src/main/java/dev/openrs2/deob/ast/transform/NewInstanceTransformer.kt
package dev.openrs2.deob.ast.transform

import com.github.javaparser.ast.CompilationUnit
import com.github.javaparser.ast.expr.MethodCallExpr
import dev.openrs2.deob.ast.util.isClass
import dev.openrs2.deob.ast.util.walk
import javax.inject.Singleton

@Singleton
class NewInstanceTransformer : Transformer() {
    override fun transformUnit(
        units: Map<String, CompilationUnit>,
        unit: CompilationUnit
    ) {
        unit.walk { expr: MethodCallExpr ->
            if (expr.nameAsString != "newInstance") {
                return@walk
            }

            expr.scope.ifPresent { scope ->
                if (scope.calculateResolvedType().isClass()) {
                    expr.setScope(MethodCallExpr(scope.clone(), "getDeclaredConstructor"))
                }
            }
        }
    }
}
